### PR TITLE
[CDAP-21070] Introduce error details provider to get more information about exceptions from plugins

### PR DESCRIPTION
**Pull request by itsankit-google**
Tested in CDAP Sandbox:

#### GCS Source
```
Caused by: io.cdap.cdap.api.exception.WrappedStageException: Stage 'NYT Best Sellers Raw Data' encountered : io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Splitting'. Error message: xxxxx.iam.gserviceaccount.com does not have storage.objects.list access to the Google Cloud Storage bucket. Permission 'storage.objects.list' denied on resource (or it may not exist).
	at io.cdap.cdap.etl.common.ErrorDetails.handleException(ExceptionUtils.java:77)
	at io.cdap.cdap.etl.spark.io.StageTrackingInputFormat.getSplits(StageTrackingInputFormat.java:55)
	at org.apache.spark.rdd.NewHadoopRDD.getPartitions(NewHadoopRDD.scala:136)
	..................................
Caused by: io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Splitting'. Error message: xxxxxx.iam.gserviceaccount.com does not have storage.objects.list access to the Google Cloud Storage bucket. Permission 'storage.objects.list' denied on resource (or it may not exist).
	at io.cdap.cdap.api.exception.ProgramFailureException$Builder.build(ProgramFailureException.java:186)
	at io.cdap.cdap.api.exception.ErrorUtils.getProgramFailureException(ErrorUtils.java:161)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getProgramFailureException(GCPErrorDetailsProvider.java:80)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getExceptionDetails(GCPErrorDetailsProvider.java:52)
	at io.cdap.cdap.etl.common.ErrorDetails.handleException(ExceptionUtils.java:75)
	... 66 common frames omitted
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
GET https://storage.googleapis.com/storage/v1/b/cdf_example/o?delimiter=/&fields=items(bucket,name,size,updated),prefixes,nextPageToken&includeTrailingDelimiter=true&maxResults=5000
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "message" : "xxxxxx.iam.gserviceaccount.com does not have storage.objects.list access to the Google Cloud Storage bucket. Permission 'storage.objects.list' denied on resource (or it may not exist).",
    "reason" : "forbidden"
  } ],
  "message" : "xxxxx.iam.gserviceaccount.com does not have storage.objects.list access to the Google Cloud Storage bucket. Permission 'storage.objects.list' denied on resource (or it may not exist)."
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
```
#### GCS Sink
```
Caused by: io.cdap.cdap.api.exception.WrappedStageException: Stage 'GCS2' encountered : io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Validating Output Specs'. Error message: xxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
	at io.cdap.cdap.etl.common.ExceptionUtils.handleException(ExceptionUtils.java:77)
	at io.cdap.cdap.etl.spark.io.StageTrackingOutputFormat.checkOutputSpecs(StageTrackingOutputFormat.java:67)
	at io.cdap.cdap.etl.common.output.MultiOutputFormat.checkOutputSpecs(MultiOutputFormat.java:114)
	.......................................
Caused by: io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Validating Output Specs'. Error message: xxxxxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
	at io.cdap.cdap.api.exception.ProgramFailureException$Builder.build(ProgramFailureException.java:186)
	at io.cdap.cdap.api.exception.ErrorUtils.getProgramFailureException(ErrorUtils.java:161)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getProgramFailureException(GCPErrorDetailsProvider.java:80)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getExceptionDetails(GCPErrorDetailsProvider.java:52)
	at io.cdap.cdap.etl.common.ErrorDetails.handleException(ExceptionUtils.java:75)
	... 47 common frames omitted
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
GET https://storage.googleapis.com/storage/v1/b/cdf_example/o/2024-10-15-08-53?fields=bucket,name,timeCreated,updated,generation,metageneration,size,contentType,contentEncoding,md5Hash,crc32c,metadata
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "message" : "xxxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).",
    "reason" : "forbidden"
  } ],
  "message" : "xxxxxxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist)."
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:428)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
```
#### GCS Multi Sink
```
Caused by: io.cdap.cdap.api.exception.WrappedStageException: Stage 'GCS Multi File' encountered : io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Committing'. Error message: xxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
	at io.cdap.cdap.etl.common.ExceptionUtils.handleException(ExceptionUtils.java:77)
	at io.cdap.cdap.etl.spark.io.StageTrackingOutputCommitter.setupJob(StageTrackingOutputCommitter.java:57)
	at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.setupJob(HadoopMapReduceCommitProtocol.scala:188)
	...............
Caused by: io.cdap.cdap.api.exception.ProgramFailureException: Error occurred in the phase: 'Committing'. Error message: xxxxxx.iam.gserviceaccount.com does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
	at io.cdap.cdap.api.exception.ProgramFailureException$Builder.build(ProgramFailureException.java:186)
	at io.cdap.cdap.api.exception.ErrorUtils.getProgramFailureException(ErrorUtils.java:161)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getProgramFailureException(GCPErrorDetailsProvider.java:80)
	at io.cdap.plugin.gcp.common.GCPErrorDetailsProvider.getExceptionDetails(GCPErrorDetailsProvider.java:52)
	at io.cdap.cdap.etl.common.ErrorDetails.handleException(ExceptionUtils.java:75)
	... 47 common frames omitted
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: null
	at com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createJsonResponseException(GoogleCloudStorageExceptions.java:89)
	at com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl$9.onFailure(GoogleCloudStorageImpl.java:1944)
	at com.google.cloud.hadoop.gcsio.BatchHelper.execute(BatchHelper.java:199)
	at com.google.cloud.hadoop.gcsio.BatchHelper.lambda$queue$0(BatchHelper.java:178)
	
````

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSinkContext.java [hunk 1]
```diff
@@ -38,4 +38,13 @@ public interface BatchSinkContext extends BatchContext {
    * @return a boolean value which indicates the pipeline is running in preview mode.
    */
   boolean isPreviewEnabled();
+
+  /**
+   * Overrides the error details provider class name.
+   *
+   * @param errorDetailsProviderClassName the error details provider class name.
+   */
+  default void setErrorDetailsProvider(String errorDetailsProviderClassName) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSinkContext.java]: is there a reason we only want it in the source/sink instead of for everything?

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSinkContext.java]: it's also more extensible if this is a class (for ex: ErrorDetailsProviderSpec) that has a getClassName() method. That way, if we find we need to be able to pass more information like configuration, this method signature doesn't need to change.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Interface for providing error details.
+ *
+ * <p>
+ * Implementations of this interface can be used to provide more detailed error information
+ * for exceptions that occur within the code using {@link ProgramFailureException}.
+ * </p>
+ */
+public interface ErrorDetailsProvider {
+  public static final String EXCEPTION_DETAILS_CLASS_NAME_KEY =
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: this constant is unrelated to the interface. It's an implementation detail of the inputformat/outputformat and should be defined somewhere there instead of in the api

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * the License.
+ */
+
+/**
+ * Enum representing the different phases of a stage where error can occur.
+ */
+package io.cdap.cdap.etl.api.exception;
+
+public enum ErrorPhase {
+  GETTING_SPLITS( "Getting Splits"),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java]: nit: extra space in front of `"`

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   * Returns a {@link RuntimeException} that wraps the given exception
+   * with more detailed information.
+   *
+   * @param e the exception to wrap.
+   * @param phase the phase of the stage where the exception occurred.
+   * @return {@link ProgramFailureException} that wraps the given exception
+   * with more detailed information.
+   */
+  @Nullable
+  ProgramFailureException getExceptionDetails(Exception e, ErrorPhase phase);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: It's more extensible to take an ErrorContext class instead of ErrorPhase, where ErrorContext starts off with a single method to get the phase. This allows us to add methods to the context in the future without changing this method signature.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Enum representing the different phases of a stage where error can occur.
+ */
+package io.cdap.cdap.etl.api.exception;
+
+public enum ErrorPhase {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java:22]: nit: 'phase' is used elsewhere in the code to refer to sections/programs within the workflow. It would be better to use some other name. Not sure what would be better here though, maybe 'operation'?

I don't have a strong opinion here though, if we think this name is good enough just be sure to add javadocs.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+   * Handles the given exception, wrapping it in a {@link WrappedStageException}.
+   *
+   * @param e the exception to handle.
+   * @param stageName the name of the stage where the exception occurred.
+   * @param errorDetailsProvider the error details provider.
+   * @param phase the phase of the stage where the exception occurred.
+   * @return the wrapped stage exception.
+   */
+  public static WrappedStageException handleException(Exception e, String stageName,
+    @Nullable ErrorDetailsProvider errorDetailsProvider, @Nullable ErrorPhase phase) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: It's better to avoid nullable when possible. It doesn't seem like the ErrorPhase should ever be null, and we can have a default instance for the ErrorDetailsProvider which is used instead of passing null here.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * Implementations of this interface can be used to provide more detailed error information
+ * for exceptions that occur within the code using {@link ProgramFailureException}.
+ * </p>
+ */
+public interface ErrorDetailsProvider {
+  public static final String EXCEPTION_DETAILS_CLASS_NAME_KEY =
+    "io.cdap.pipeline.error.details.provider.classname";
+
+  /**
+   * Returns a {@link RuntimeException} that wraps the given exception
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: RuntimeException -> ProgramFailureException

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      conf.get(ErrorDetailsProvider.EXCEPTION_DETAILS_CLASS_NAME_KEY);
+    if (errorDetailsProviderClassName == null) {
+      return null;
+    }
+    try {
+      return (ErrorDetailsProvider) conf.getClassLoader()
+        .loadClass(errorDetailsProviderClassName)
+        .newInstance();
+    } catch (Exception e) {
+      throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategoryEnum.PLUGIN),
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: when would this happen? Wondering if it makes sense to log a warning and return some default provider rather than fail here.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+   *
+   * @param conf the configuration to get the error details provider from.
+   * @return the error details provider.
+   */
+  @Nullable
+  public static ErrorDetailsProvider getErrorDetailsProvider(Configuration conf) {
+    String errorDetailsProviderClassName =
+      conf.get(ErrorDetailsProvider.EXCEPTION_DETAILS_CLASS_NAME_KEY);
+    if (errorDetailsProviderClassName == null) {
+      return null;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: would be good to implement some default instance that returns null for each exception and return that here. Then the method doesn't need to be nullable.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java]: nit: less verbose to change these to SPLITTING, READING, WRITING, COMMITTING

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/StageTrackingOutputFormat.java [hunk 10]
```diff
@@ -39,47 +41,53 @@ public class StageTrackingOutputFormat<K, V> extends DelegatingOutputFormat<K, V
...
   public RecordWriter<K, V> getRecordWriter(TaskAttemptContext context) {
     OutputFormat<K, V> delegate = getDelegate(context.getConfiguration());
+    Configuration conf = context.getConfiguration();
     try {
       // Spark already emitting bytes written metrics for file base output,
       // hence we don't want to double count
       if (delegate instanceof FileOutputFormat) {
         return new StageTrackingRecordWriter<>(delegate.getRecordWriter(context),
-          getStageName(context.getConfiguration()));
+          getStageName(conf), ExceptionUtils.getErrorDetailsProvider(conf));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/StageTrackingOutputFormat.java]: If it can throw an exception, its better to create the ErrorDetailsProvider outside of this try-catch. Otherwise an error instantiating it will get treated like an error writing records.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSinkContext.java (hunk 1 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSinkContext.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java (hunk 3 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java (hunk 4 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java (hunk 5 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorPhase.java:22]: ack, added a javadoc comment.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java (hunk 6 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java (hunk 7 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java (hunk 8 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: ack, done.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java (hunk 9 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: ack, done.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/StageTrackingOutputFormat.java (hunk 10 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/io/StageTrackingOutputFormat.java]: since it does not throw an exception now, kept the change.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java (hunk 2 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/exception/ErrorDetailsProvider.java]: ack, moved to `ExceptionUtils` class.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import io.cdap.cdap.etl.api.exception.ErrorPhase;
+import io.cdap.cdap.etl.api.exception.NoopErrorDetailsProvider;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for handling exceptions.
+ */
+public class ExceptionUtils {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: it's better to name utility classes more specific, otherwise they tend to accumulate a bunch of loosely related logic.

For example, ErrorDetails instead of ExceptionUtils.

**Review by albertshau — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java (hunk 11 above)
Comment (itsankit-google) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/ExceptionUtils.java]: ack, done.
